File: TFTP Server/python/tftp_server.py
#
# IPWorks 2022 Python Edition - Demo Application
#
# Copyright (c) 2023 /n software inc. - All rights reserved. - www.nsoftware.com
#

import sys
import string
from ipworks import *

input = sys.hexversion<0x03000000 and raw_input or input

print("********************************************************************************\n")
print("* This demo shows how to set up a TFTP server using the TFTPServer component.  *\n")
print("********************************************************************************\n")

try:
  server = TFTPServer()

  def fireError(e):
    print(e.description)

  def fireLog(e):
    print("Log: " + e.message)

  def fireConnected(e):
    print("%d: connected." % e.connection_id)

  def fireConnectionRequest(e):
    print(e.remote_host + ": attempting to connect.")

  def fireDisconnected(e):
    print("%d: disconnected." % e.connection_id)

  def fireStartTransfer(e):
    print("%d: started a transfer." % e.connection_id)

  def fireEndTransfer(e):
    print("%d: transfer complete." % e.connection_id)

  def fireTransfer(e):
    print("%d: transferring data." % e.connection_id)

  server.on_connected = fireConnected
  server.on_connection_request = fireConnectionRequest
  server.on_disconnected = fireDisconnected
  server.on_start_transfer = fireStartTransfer
  server.on_transfer = fireTransfer
  server.on_end_transfer = fireEndTransfer
  server.on_error = fireError

  server.set_local_dir(input("Local Directory: "))

  server.start_listening()

  print("Listening...")
  print("Press Ctrl-C to shutdown")

  while True:
    server.do_events()

except IPWorksTftpserverError as e:
  print("ERROR: " + e.message)

except KeyboardInterrupt:
  print("Shutdown requested...exiting")
  server.stop_listening()

File: TFTP Server/cpp/tftpserver.cpp
/*
 * IPWorks 2022 C++ Edition - Demo Application
 *
 * Copyright (c) 2023 /n software inc. - All rights reserved. - www.nsoftware.com
 *
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworks.h"

class MyTFTPServer : public TFTPServer
{
public: 
    virtual int FireConnectionRequest(TFTPServerConnectionRequestEventParams *e) {
        printf("[%s:%d]: Attempting to connect.\n", e->RemoteHost, e->RemotePort);
        return 0;
    }

    virtual int FireConnected(TFTPServerConnectedEventParams* e) {
        printf("[%d]: Connected.\n", e->ConnectionId);
        return 0;
    }

    virtual int FireStartTransfer(TFTPServerStartTransferEventParams* e) {
        if (e->Direction == 0) {
            printf("[%d]: Upload started.\n", e->ConnectionId);
        }
        else {
            printf("[%d]: Download started.\n", e->ConnectionId);
        }
        return 0;
    }

    virtual int FireTransfer(TFTPServerTransferEventParams *e) {
        printf("Transferring data.\n");
        return 0;
    }

    virtual int FireEndTransfer(TFTPServerEndTransferEventParams* e) {
        if (e->Direction == 0) {
            printf("[%d]: Upload complete.\n", e->ConnectionId);
        }
        else {
            printf("[%d]: Download complete.\n", e->ConnectionId);
        }
        return 0;
    }

    virtual int FireDisconnected(TFTPServerDisconnectedEventParams* e) {
        printf("[%d]: Disconnected.\n", e->ConnectionId);
        return 0;
    }
};

int main(int argc, char** argv)
{
    printf("This demo shows how to use the TFTPServer Component to create a simple TFTP Server.\n");
    printf("Press Ctrl+C to shutdown the server.\n");

    if (argc != 2) {
        fprintf(stderr, "usage: tftpserver path/to/localdir\n");
        fprintf(stderr, "\nExample: tftpserver C:\\temp \n\n");
        printf("Press enter to continue.");
        getchar();
    }
    else {
        MyTFTPServer server;

        server.SetLocalDir(argv[1]);

        server.StartListening();

        printf("Server is listening on port: %d\n", server.GetLocalPort());

        while (true) {
            server.DoEvents();
        }
        exit(0);
    }
}





